### ISSUE TITLE
Default uBO in addition to certain Firefox settings has the potential to crash not only the web browser but the entire Operating System

### ISSUE THREAD

**Issue by Pxartist** [uBlockOrigin/uBlock-issues#3468]
### Prerequisites

- [X] I verified that this is not a filter list issue. Report any issues with filter lists or broken website functionality in the [uAssets issue tracker](https://github.com/uBlockOrigin/uAssets/issues).
- [X] This is **NOT** a [_YouTube_](https://www.github.com/uBlockOrigin/uAssets/issues/20586), [_Facebook_](https://www.github.com/uBlockOrigin/uAssets/issues/3367) or [_Twitch_](https://www.github.com/uBlockOrigin/uAssets/issues/5184) report. _These sites MUST be reported by clicking their **respective links**_.
- [X] This is not a support issue or a question. For support, questions, or help, visit [/r/uBlockOrigin](https://www.reddit.com/r/uBlockOrigin/).
- [X] I performed a [cursory search of the issue tracker](https://github.com/uBlockOrigin/uBlock-issues/issues?q=is%3Aissue) to avoid opening a duplicate issue.
- [X] The issue is not present after disabling uBO in the browser.
- [X] I checked the [documentation](https://github.com/gorhill/uBlock/wiki) to understand that the issue I am reporting is not normal behavior.

### I tried to reproduce the issue when...

- [X] uBO is the only extension.
- [X] uBO uses default lists and settings.
- [X] using a new, unmodified browser profile.

### Description

[https://www.reddit.com/r/uBlockOrigin/comments/1h0f3wg/ublockorigin_can_cause_the_black_screen_of_death/](https://www.reddit.com/r/uBlockOrigin/comments/1h0f3wg/ublockorigin_can_cause_the_black_screen_of_death/)

### A specific URL where the issue occurs.

```yaml
https://www.wealthfront.com/

Possibly more websites
```

### Steps to Reproduce

1.    Create a new Firefox Profile

   2.  Create prefs.js in the new profile with the contents: https://hastebin.skyra.pw/eyazuwagun.less

3.     Launch Firefox using the new profile and download uBO from https://addons.mozilla.org/en-US/firefox/addon/ublock-origin/

  4.   Visit https://www.wealthfront.com/ to turn Firefox into Google Chrome while acknowledging the following terms:

      a.  You will not hold me responsible for any lost work or damages for your system entering a black screen or irrecoverable state.

        b. Closing the offending tab might not be enough to save your session. I have "successfully" crashed my system 3 more times trying to write this post while losing even more work.

            c. Closing the tab may or may not prevent excessive consumption. On my system, I estimate that the rate of memory consumption to be from 3 - 5 GB / second. The memory consumption will continue for a few moments even after the tab is closed.

            d. A rebound / refractory period may occur after 1 or 2 minutes in which excessive memory consumption will occur again and not stop until the system hangs, crashes, or enter a lobotomized state with many serious malfunctions requiring a complete system restart.

### Expected behavior

My system not crashing in 5 minutes when I visit the webpage. At the very least, the tab should automatically crash or kill itself.

### Actual behavior

My system does one of the following:

1. Briefly consume up to 100 GB of RAM in less than a minute just for visiting the website.
2. Briefly consume up to 100 GB of RAM just for visiting the website. After 1 - 2 minutes, RAM will begin to be consumed at a rapid rate until there is none left and the OS will either crash or enter a lobotomized state.
3. Consume all RAM and crash.

### uBO version

 1.61.2

### Browser name and version

Firefox 132.0.2

### Operating System and version

Windows 11 Pro 23H2

**Comment by stephenhawk8054** [comment]
> At the very least, the tab should automatically crash or kill itself.

This should be handled by the browsers.

**Comment by gorhill** [comment]
I can't reproduce with default settings and default browser settings.

From the Reddit thread you posted, it appears the site is firing a firehose of `navigator.sendBeacon()` requests, and in such case this can lead to Firefox not garbage-collecting its memory fast enough, there was an issue on Bugzilla about this.

> Create prefs.js in the new profile with the contents: https://hastebin.skyra.pw/eyazuwagun.less

You will have to figure out what in there is causing your issue. In your Reddit post you say _"I don't know squat about computers"_, and in such case I advise you don't heavily modify browser default settings.

**Comment by gorhill** [comment]
I created a new profile with `https://hastebin.skyra.pw/eyazuwagun.less`, installed uBO 1.61.2, and still cannot reproduce your reported issue.

**Comment by gorhill** [comment]
@gwarser Found the specific config which triggers the issue:

> user_pref("beacon.enabled", false);

I could reproduce this way (somehow the `prefs.js` file didn't work.)

From the dev network tab, I could see huge amount of network requests being fired, so this is the same sort of issue as the one that was filed in Bugzilla, which is that the garbage collection doesn't occur fast enough when a content blocker is installed and a site fires network requests in a loop non-stop.

This happens with any content blocker which blocks the network requests the site is trying to fire, not just uBO.

**Comment by Pxartist** [comment]
> this is the same sort of issue as the one that was filed in Bugzilla

Could you provide the link to the report? I don't see any reports that pertain to system crashing or garbage collection not being fast enough when looking up "beacon.enabled".


Would you also be able to explain why a combination of https://www.wealthfront.com/, default uBO, and 

    user_pref("beacon.enabled", false);
is able to have such disastrous consequences when visiting the website? I have visited many other websites without this issue ever occurring. If anything, I am surprised that I'm still able to use Youtube.com, which uBO regularly blocks >1k entities, without my system crashing.

**Comment by gorhill** [comment]
> Could you provide the link to the report?

I can't find it, but I know it's in there somewhere in <https://bugzilla.mozilla.org/query.cgi?format=specific>. I think the issue was closed as won't fix.

It's essentially a site firing network requests non stop in a loop and at such a fast rate that the browser's garbage collector was not being fast enough to free up the transient memory consumed by the webRequest API when a content blocker is used.

Chromium was also affected but its garbage collector was able kicking in so it was less affected.

**Comment by Pxartist** [comment]
Is the site firing so many network requests because of uBO? Can anything be done on uBO's end to prevent such an issue from occurring?

Shouldn't there be safeguards already implemented to prevent such an issue from occurring to begin with? I don't see why other anti-content blocker websites couldn't implement such tactics to eliminate the usage of content blockers.

**Comment by gorhill** [comment]
> Is the site firing so many network requests because of uBO?

Not because of uBO specifically, as already pointed out, it happens with other content blockers. The site has a defect causing firehose of network requests, and Firefox's garbage collection does not kick in. That is the cause of the issue.

> I don't see why other anti-content blocker websites couldn't implement such tactics to eliminate the usage of content blockers.

Yes, sites are free to sabotage themselves when you visit them. I am pretty sure this will eliminate visits to their site before it eliminate content blockers.

**Comment by gorhill** [comment]
I think we explained well the issue, I just wish I could track the bugzilla entry with same behavior as reported here -- the Firefox devs did point out that the garbage collector was not kicking in in due time.

At this point, I can only advise that if you do not trust our findings, best is to uninstall uBO and find something else you trust.

**Comment by gorhill** [comment]
> maybe something to do with some 3rd party script ?

The issue is in `https://www.wealthfront.com/_next/static/chunks/pages/_app-c02914ec03b315c3.js`, search for call to `R.trackESP(e, t, n, o)`.

The site code is flawed. If `navigator.sendBeacon` is `undefined`, it will fallback to using `XMLHttpRequest`, and when `XMLHttpRequest` fails (because the request is blocked by uBO, or any content blocker, or any other ways including from outside the browser), the site's code will repeatedly try to send the error condition to its server, which will again fail, etc forever.

Can reproduce the same issue in Chromium by setting `navigator.sendBeacon` to `undefined` after setting a breakpoint to above call, set `navigator.sendBeacon` to `undefined`, remove the breakpoint, and see memory usage growing in the browser's Task manager.

Anyway, a site flaw which becomes visible because changing browser configurations. Though the site readies itself to deal with `navigator.sendBeacon` not being supported, its fallback code is flawed. If it want to fallback to `XMLHttpRequest` when `navigator.sendBeacon` is not available, then it should make its fallback code behave the same: `navigator.sendBeacon` never returns a response from the server, it's a one-way request, so it should not care about `XMLHttpRequest` failing to return a response.

**Comment by gwarser** [comment]
@gorhill maybe this: https://github.com/uBlockOrigin/uAssets/issues/12797#issuecomment-1103859376

**Comment by gorhill** [comment]
Right, there was this one too: https://bugzilla.mozilla.org/show_bug.cgi?id=1504623

But as far I can tell, it's first the site's own flawed code, even garbage collection won't help.

**Comment by stephenhawk8054** [comment]
> I don't see why other anti-content blocker websites couldn't implement such tactics to eliminate the usage of content blockers.

If a site wants to actively implement "anti-blocker tactics", there are thousands of ways to do that already, and we deal with those sites daily. Sorry, there are no assured "fallbacks" if a site deliberately circumvents blockers, we know that because we deal with those kinds of sites all the time. Your "fallbacks" will need another "fallback" again, again and again.

As said above, if you don't trust our findings, feel free to use other products that meet your expectations.

**Comment by stephenhawk8054** [comment]
@gorhill Another solution that works for me is

```adb
wealthfront.com##+js(set, navigator.sendBeacon, noopFunc)
```

Should we address it in resource abuse list?

**Comment by gorhill** [comment]
We shouldn't address it. Firefox does warn prominently about changing anything in `about:config`, and the case here is an example of bad side effect of doing so. Also by default uBO already blocks 3rd-party `beacon` requests, and some 1st-party as seen here, there is no point blanking `sendBeacon`.

Our filter lists are for default uBO settings and reasonable browser configurations. Blanking `sendBeacon` is bound to causes more issues elsewhere, so we should not consider this reasonable config change -- [that method was added more than 10 years ago, so sites reasonably expects it](https://caniuse.com/?search=sendBeacon).

**Comment by Pxartist** [comment]
> If a site wants to actively implement "anti-blocker tactics", there are thousands of ways to do that already

Could you provide examples?

**Comment by Pxartist** [comment]
> The site code is flawed. If `navigator.sendBeacon` is `undefined`, it will fallback to using `XMLHttpRequest`, and when `XMLHttpRequest` fails (because the request is blocked by uBO, or any content blocker, or any other ways including from outside the browser), the site's code will repeatedly try to send the error condition to its server, which will again fail, etc forever.

So the site's catch all solution with dealing with navigator.sendBeacon works with the expectation of always succeeding and does not anticipate to fail due to the presence of a content blocker, such as uBO. The unanticipated failure causes Firefox to quickly allocate memory nonstop until a system failure occurs.



> [that method was added more than 10 years ago, so sites reasonably expect it](https://caniuse.com/?search=sendBeacon).

And https://www.wealthfront.com/ is able to handle navigator.sendBeacon with the value of 'undefined' without the presence of a content blocker by implementing a catchall solution that was designed with the expectation of always succeeding. In that case, it is not my fault for having 

    user_pref("beacon.enabled", false);

The unexpected consequence of the scenario resulting in a system crash is a combination of the following:

- https://www.wealthfront.com/ having a catchall solution when dealing with an 'undefined' navigator.sendbeacon that can cause Firefox to allocate memory nonstop to the point of system failure if the catchall fails in any way.
- uBO playing an active part in causing https://www.wealthfront.com/ catchall solution to fail by blocking or interfering with the catchall solution's execution.
- Firefox not recognizing the need to stop allocating memory or killing the tab before system failure occurs. While it could be argued that a better garbage collector would prevent the system crash would help, a browser tab making Firefox allocate 3 - 5 GB of RAM per second is definitely abnormal behavior and should not be allowed to continue.


================================================================================

While my scenario still requires the presence of user_pref("beacon.enabled", false);, **I do not see why a website, such as Youtube.com, wouldn't implement a way to request to serve undesirable content nonstop and make Firefox crash or even possibly bring down the user's computer because the undesirable content fails to be properly served due to the intervention of a content blocker.**

**Comment by Pxartist** [comment]
> @Pxartist https://github.com/uBlockOrigin/uAssets/issues

Link me examples of the system crashing or at the very least the web browser crashing.

**Comment by stephenhawk8054** [comment]
I said

> If a site wants to actively implement "anti-blocker tactics", there are thousands of ways to do that already

It doesn't mean that all of the methods require stressing the machine. That's it.

If you want to experience how the websites do the anti-blockers, try contributing and solving the issues in uAssets, then you'll understand.

And as I said, there are no assured "fallbacks" as you wish, because your "fallbacks" will need other "fallbacks" repeatedly every day. Managing machines' resources is the job of the browsers, not extensions. They don't have enough privileges to do that. If you want extensions to do that job, find other products that meet your expectations.
